sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/DoNotUseCollectionInItsOwnMethodCalls.cs

```diff
@@ -0,0 +1,182 @@
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(rule);
+
+        private static readonly ISet<string> trackedMethodNames = ImmutableHashSet.Create("AddRange", "Concat",
+            "Union", "Except", "Intersect", "SequenceEqual", "UnionWith", "ExceptWith");
+
+        protected override void Initialize(SonarAnalysisContext context)
+        {
+            context.RegisterSyntaxNodeActionInNonGenerated(
+                c =>
+                {
+                    var invocation = (InvocationExpressionSyntax)c.Node;
+                    var invocationExpressionString = invocation.Expression.ToString();
+
+                    if (!trackedMethodNames.Any(method => invocationExpressionString.EndsWith(method,
+                        StringComparison.Ordinal)))
+                    {
+                        return;
+                    }
+
+                    var operands = GetOperandsToCheckIfTrackedMethod(invocation, c.SemanticModel);
+                    if (operands != null &&
+                        EquivalenceChecker.AreEquivalent(operands.Left, operands.Right))
+                    {
+                        c.ReportDiagnostic(Diagnostic.Create(rule, operands.Left.GetLocation(),
+                            additionalLocations: new[] { operands.Right.GetLocation() },
+                            messageArgs: new[] { operands.Right.ToString(), operands.ErrorMessage }));
+                    }
+                }, SyntaxKind.InvocationExpression);
+        }
+
+        private static OperandsToCheck GetOperandsToCheckIfTrackedMethod(InvocationExpressionSyntax invocation,
+            SemanticModel model)
+        {
+            var methodSymbol = model.GetSymbolInfo(invocation).Symbol as IMethodSymbol;
+
+            bool isEnumerableMethod;
+            var message = ProcessIssueMessageFromMethod(methodSymbol, out isEnumerableMethod);
+            if (message == null)
+            {
+                return null;
+            }
+
+            if (isEnumerableMethod && methodSymbol.MethodKind == MethodKind.Ordinary)
+            {
+                return new OperandsToCheck(invocation.ArgumentList.Arguments[0].Expression,
+                    invocation.ArgumentList.Arguments[1].Expression, message);
+            }
+
+            var invokingExpression = (invocation.Expression as MemberAccessExpressionSyntax)?.Expression;
+            if (invokingExpression != null)
+            {
+                return new OperandsToCheck(invokingExpression, invocation.ArgumentList.Arguments[0].Expression,
+                    message);
+            }
+
+            return null;
+        }
+
+        private static string ProcessIssueMessageFromMethod(IMethodSymbol methodSymbol, out bool isEnumerableMethod)
+        {
+            isEnumerableMethod = true;
+
+            if (methodSymbol.IsEnumerableConcat())
```

Comment:
They are not for IsISetExceptWith and IsEnumerableExcept. But I noticed it's because isEnumerableMethod is an out parameter - so ignore this one as well :/